Allow the no_stack_protector attribute on local variables

cooperp opened the pull request:
This attribute is currently only permitted on functions, but sometimes it can be useful to opt out a specific local variable.  That way the function can still get a stack protector if other locals require it, but locals which are opted out won't have the overhead of a stack protector if not needed.

This is paired with an LLVM change which has already landed, which uses the new metadata to opt specific alloca instructions out of requiring a stack protector.

AaronBallman reviewed (commented):
The changes should also come with a release note in `clang/docs/ReleaseNotes.rst` and `clang/include/clang/Basic/AttrDocs.td` needs to have `NoStackProtectorDocs` updated as well.

cooperp commented:
> Ope, I missed this comment thread when I was leaving review comments. Picking a new name for the attribute would make sense to me in this case given the landscape.

Updated to use the suggested name from @efriedma-quic - `stack_protector_ignore`.  And to add docs, release notes, etc.

cooperp commented on clang/test/CodeGen/stack-protector-vars.cpp:
```diff
@@ -0,0 +1,38 @@
+}
+
+// CHECK: define {{.*}} @_Z5test3v
+// CHECK-NOT: %b = alloca [1000 x i8], align {{.*}}, !stack-protector
```
Oh sorry silly copy/paste of the check line. I'd expect the alloca in test3 to have the attribute. I just checked for the wrong thing.

cooperp replied:
Thanks!  Yes thats much better. Not sure why i didn't think of that before.

AaronBallman reviewed (commented):
Thank you! I think this is missing some Sema tests for things like applying the attribute to a non-local variable, a static local variable, a non-variable, and being given an argument list. I'd expect each of those to end up being diagnosed, so we want to verify the diagnostics are issued and reasonable.

cooperp committed:
Allow the no_stack_protector attribute on local variables.

This attribute is currently only permitted on functions, but sometimes it can
be useful to opt out a specific local variable.  That way the function can still
get a stack protector if other locals require it, but locals which are opted out
won't have the overhead of a stack protector if not needed.

cooperp commented:
> Thank you! I think this is missing some Sema tests for things like applying the attribute to a non-local variable, a static local variable, a non-variable, and being given an argument list. I'd expect each of those to end up being diagnosed, so we want to verify the diagnostics are issued and reasonable.

Good idea!  Added `clang/test/Sema/stack_protector_ignore.c` with the suggested cases. Not sure if i got what you'd expect for the "non-variable" case but i did get a warning when applying this to an add expression.

cooperp commented on clang/test/CodeGen/stack-protector-vars.cpp:
```diff
@@ -0,0 +1,68 @@
+// RUN: %clang_cc1 -triple x86_64-apple-darwin -emit-llvm -o - %s | FileCheck %s
```
There's 2 cases.  So if you have:

    void func2(void) {
        __attribute__((stack_protector_ignore)) int var;
    }

then if you do `-fstack-protector` you'll get the `ssp` attribute on the function, and the `stack-protector=0` attribute on the AllocaInst.

If you do `-fstack-protector-all` then you get `sspreq` on the function, and still the `stack-protector=0` attribute on the AllocaInst.

I could certainly have a test for these cases, but i wonder if in the `-fstack-protector-all` case whether it would be reasonable for the `CGDecl.cpp` code i'm adding to instead skip the `stack-protector=0` on the AllocaInst?  That is, only add the `stack-protector=0` if using `-fstack-protector` not `-fstack-protector-all`? That would give me something good to test for.

What do you think?

cooperp replied:
Realized after i wrote the above that it's reasonable to still emit the attribute on the alloca even if the function has `sspreq` (stack-protector-all). If someone wanted to inline this cross module then I think its reasonable for the alloca to keep its attribute post-inlining, such that it opts out of the stack protector heuristic (if any) in its final function.

Saying all that, i've updated the codegen test to be invoked a few different times with different command line flags to sure we always get the attributes on the alloca's as expected.

cooperp commented on clang/include/clang/Basic/AttrDocs.td:
```diff
@@ -5736,6 +5736,17 @@ option.
+variable when determining if a function should use a stack protector.
+
+If ``-fstack-protector-all`` is specified then the given function will still
+use a stack protector, even if some/all variables have this attribute.
```
I was looking to check for `-fstack-protector-all` in `CGDecl.cpp` where i've made the changes in this PR.  Then at that point emit a warning too if the new ignore attribute is present.

Unfortunately running in to an ordering issue. The stack protector flags are checked in `CodeGenModule::SetLLVMFunctionAttributesForDefinition()` but that is after i'm emitting the attribute in `CodeGenFunction::EmitAutoVarAlloca()`.

I could potentially factor out the stack protector logic in to a helper, eg:

    clang::LangOptions::StackProtectorMode CodeGenModule::StackProtectorMode(const Decl *D) const;

then call that helper from `SetLLVMFunctionAttributesForDefinition()` to handle the current logic of attaching attributes, but also call it from `EmitAutoVarAlloca()` to see if we should emit this new suggested warning.

Alternatively, if this is totally the wrong direction just let me know @AaronBallman. I don't know if this is the kind of thing that something like Sema should handle instead as i'm more familiar with codegen and IR than the rest of clang.

cooperp replied:
I've added `CodeGenModule::StackProtectorAttribute(const Decl *D)` so that i could get the attribute kind (if any) for a given decl.  That is then used from `SetLLVMFunctionAttributesForDefinition()` as it is today.

I also use it from `EmitAutoVarAlloca()` such that when the `stack_protector_ignore` attribute is on a variable, you can get a warning such as:

    'stack_protector_ignore' attribute ignored due to '-fstack-protector-all' option

AaronBallman reviewed (commented):
Test failures are relevant, I left a comment where I think the issue is.

cooperp commented on clang/include/clang/Basic/DiagnosticCommonKinds.td:
```diff
@@ -312,6 +312,10 @@ def warn_stack_clash_protection_inline_asm : Warning<
 
+def warn_stack_protection_ignore_attribute : Warning<
+  "'stack_protector_ignore' attribute ignored due to "
+  "'-fstack-protector-all' option">;
```
Thanks @AaronBallman!  I applied that fix and it passes locally with `check-clang`

AaronBallman reviewed (approved):
Thank you! LGTM assuming @efriedma-quic is okay with the changes to codegen.

efriedma-quic reviewed (approved):
LGTM
